### ISSUE TITLE
Mach_Jet with AMR

### ISSUE THREAD

**Issue by ChrisDeGrendele** [AMReX-Astro/MAESTROeX#203]
`MAESTROeX/Exec/test_problems/mach_jet` when run with AMR segfaults. On a second run, it gives the following error:

```
Assertion `nghost.allLE(nGrowVect())' failed, file "/home/cdegrendele/codes/amrex//Src/Base/AMReX_FabArray.H", line 1875, Msg: "FillBoundary: asked to fill more ghost cells than we have" !!!
SIGABRT
```
When run with amr.max_level > 1 it always segfaults.

To reproduce: 
Compile mach_jet setup as is with the current main branch: `make`
Run with: `./Maestro2d.gnu.MPI.ex inputs_2d amr.max_level=1`

**Comment by ChrisDeGrendele** [comment]
Trying to figure out if this is an issue in general or problem specific: 

I tried `test_problems/rt`, `test_problems/double_bubble`, `science/toy_convect`  with AMR and got this same error with `amr.max_level =1`. 

`test_problems/incomp_shear_jet` has the following error:

```
0::Assertion `getBDKey() == m_bdkey' failed, file "/home/cdegrendele/codes/amrex//Src/Base/AMReX_FabArrayBase.cpp", line 1435 !!!
SIGABRT
```

`test_problems/reacting_bubble`, `test_problems/test_convect`run fine with AMR.

**Comment by ChrisDeGrendele** [comment]
Some progress - 

Both of the below blocks start a loop over a geometry[lev] starting at lev =1 (the first level of refinement, not the coursest level). MAESTROeX creates the grids as we need them, but the geometry is created off the bat in [Maestrosetup](https://github.com/AMReX-Astro/MAESTROeX/blob/9724da472d74210f966a6f8d3b21d6768aafacab/Source/MaestroSetup.cpp#L194) with a `max_level+1` size. 

In mach_jet and others above, we define an `amr.max_level > 1` but the tagging is not yet needed. The geometry has been created, so we can successfully loop through any geometry[lev]. But it causes problems when we assume there is actual data there. 

Below, in the first example, MAESTROeX assumes there is data in the first level, and in the second routine, AMReX assumes there is data from the size of the geometry. 

[The following is here](https://github.com/AMReX-Astro/MAESTROeX/blob/9724da472d74210f966a6f8d3b21d6768aafacab/Source/MaestroRegrid.cpp#L147)

```
 for (int lev = 1; lev <= base_geom.max_radial_level; ++lev) {

        for (MFIter mfi(sold[lev], false); mfi.isValid(); ++mfi) {
            const Box& validBox = mfi.validbox();
            // re-compute tag_array since the actual grid structure changed due to buffering
            // this is required in order to compute numdisjointchunks, r_start_coord, r_end_coord
            RetagArray(validBox, lev);
        }
    }
    ParallelDescriptor::ReduceIntMax(tag_array.dataPtr(),(base_geom.max_radial_level+1)*base_geom.nr_fine);
}
```

And the following is from [amrex/Src/LinearSolvers/MLMG/AMReX_MLLinOp.cpp](https://github.com/AMReX-Codes/amrex/blob/development/Src/LinearSolvers/MLMG/AMReX_MLLinOp.cpp#L543) where `m_num_amr_levels` is equivelent to `amr.max_level+1`
```
    for (int amrlev = 1; amrlev < m_num_amr_levels; ++amrlev)
    {
      AMREX_ASSERT_WITH_MESSAGE(m_grids[amrlev][0].coarsenable(m_amr_ref_ratio[amrlev-1]),
                                  "MLLinOp: grids not coarsenable between AMR levels");
    }
```
called [here in MAESTROeX](https://github.com/AMReX-Astro/MAESTROeX/blob/main/Source/MaestroNodalProj.cpp#L181)

1. A possible solution would be to not define the geometry as the `max_level+1` size at the beginning, but every time we need to incorporate another level of refinement, to resize all of the geometries. 
2. Another possible solution could be to make sure tag_array is only looking at defined grids and to pass slices of the geometry to AMReX. (I'm not quite sure of the syntax, but pretending this is python `a_small_geom = geom[0:2]`) and we could pass that to AMReX.


I have a meeting with @dwillcox in 10 minutes to discuss ways forward.

**Comment by ajnonaka** [comment]
Fixed with #212
